app_check/integration_test/src/integration_test.cc

```diff
@@ -821,4 +919,191 @@ TEST_F(FirebaseAppCheckTest, TestFunctionsFailure) {
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  bool received_pending_cache = false;
+  bool received_permission_denied = false;
+  firebase::firestore::ListenerRegistration registration =
+      document.AddSnapshotListener(
+          [&received_pending_cache, &received_permission_denied](
+              const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            // We first expect one call for a successful change to the cache.
+            if (error_code == firebase::firestore::kErrorNone) {
+              EXPECT_FALSE(received_pending_cache);
+              EXPECT_TRUE(result.metadata().has_pending_writes());
+              EXPECT_TRUE(result.metadata().is_from_cache());
+              received_pending_cache = true;
+            } else {
+              // We expect one call with a Permission Denied error, from the
+              // server.
+              EXPECT_FALSE(received_permission_denied);
+              EXPECT_EQ(error_code,
+                        firebase::firestore::kErrorPermissionDenied);
+              received_permission_denied = true;
+            }
+          });
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("transaction")}}),
+      "document.Set transaction", firebase::firestore::kErrorPermissionDenied);
+
+  registration.Remove();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("final")}}),
+      "document.Set final", firebase::firestore::kErrorPermissionDenied);
+
+  EXPECT_TRUE(received_pending_cache);
+  EXPECT_TRUE(received_permission_denied);
+
+  CleanupFirestore(firebase::firestore::kErrorPermissionDenied);
+}
+
+TEST_F(FirebaseAppCheckTest, TestRunTransaction) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("foo")}}),
+      "document.Set");
+
+  auto transaction_future = firestore_->RunTransaction(
+      [&](firebase::firestore::Transaction& transaction,
+          std::string&) -> firebase::firestore::Error {
+        transaction.Update(
+            document,
+            firebase::firestore::MapFieldValue{
+                {"int", firebase::firestore::FieldValue::Integer(123)}});
+        return firebase::firestore::kErrorOk;
+      });
+
+  WaitForCompletion(transaction_future, "firestore.RunTransaction");
+
+  // Confirm the updated doc is correct.
+  auto future = document.Get(firebase::firestore::Source::kServer);
+  WaitForCompletion(future, "document.Get");
+  EXPECT_THAT(future.result()->GetData(),
```

Comment:
Insert `ASSERT_NE(future.result(), nullptr)` above this line to avoid a crash if `future.result()` were ever null.

Reply:
Done